v3/src/components/graph/components/legend/categorical-legend.tsx

```diff
@@ -29,19 +29,51 @@ interface Key {
 interface Layout {
   maxWidth: number
   fullWidth: number
-  numColumns: number,
-  numRows: number,
+  numColumns: number
+  numRows: number
   columnWidth: number
 }
 
+interface DragInfo {
+  indexOfCategory: number
+  initialOffset: { x: number, y: number }
+  currentDragPosition: { x: number, y: number }
+}
+
 const keySize = 15,
   padding = 5
 
+interface LayoutData {
+  maxWidth: number
+  fullWidth: number
+  numColumns: number
+  numRows: number
+  columnWidth: number
+}
+
+const labelHeight = getStringBounds('Wy', graphVars.graphLabelFont).height
+
+const coordinatesToCatIndex = (lod: LayoutData, numCategories: number, localPoint: { x: number, y: number }) => {
+    const {x, y} = localPoint,
+      col = Math.floor(x / lod.columnWidth),
+      row = Math.floor(y / (keySize + padding)),
+      catIndex = row * lod.numColumns + col
+    return catIndex >= 0 && catIndex < numCategories ? catIndex : -1
+  },
+  catLocation = (lod: LayoutData, catData: Key[], index: number) => {
+    return {
+      x: axisGap + catData[index].column * lod.columnWidth,
+      y: /*labelHeight + */catData[index].row * (keySize + padding)
+    }
+  }
+
 export const CategoricalLegend = memo(function CategoricalLegend(
   {transform}: ICategoricalLegendProps) {
-  const dataConfiguration = useDataConfigurationContext(),
+  const transformRef = useRef(transform),
+    dataConfiguration = useDataConfigurationContext(),
     dataset = dataConfiguration?.dataset,
     layout = useGraphLayoutContext(),
+    legendBoundsRef = useRef<Bounds>(layout?.getComputedBounds('legend')),
```

Comment:
This appears to only be used in `onDragStart`. Rather than storing it in a ref and updating it on every render (line 94), it would be simpler to just call `layout?.getComputedBounds('legend')` where you need it in `onDragStart`.